packages/angular/src/tracing.ts

```
@@ -0,0 +1,208 @@
+ */
+export function TraceClassDecorator(): ClassDecorator {
+  let tracingSpan: Span;
+
+  return (target: Function) => {
+    // tslint:disable-next-line:no-unsafe-any
+    const originalOnInit = target.prototype.ngOnInit;
+    // tslint:disable-next-line:no-unsafe-any
+    target.prototype.ngOnInit = function(...args: any[]): ReturnType<typeof originalOnInit> {
+      const activeTransaction = getActiveTransaction();
+      if (activeTransaction) {
+        tracingSpan = activeTransaction.startChild({
+          description: `<${target.name}>`,
+          op: `angular.initialize`,
+        });
+      }
+      if (originalOnInit) {
+        // tslint:disable-next-line:no-unsafe-any
+        return originalOnInit.apply(this, args);
+      }
+    };
+
+    // tslint:disable-next-line:no-unsafe-any
+    const originalAfterViewInit = target.prototype.ngAfterViewInit;
+    // tslint:disable-next-line:no-unsafe-any
+    target.prototype.ngAfterViewInit = function(...args: any[]): ReturnType<typeof originalAfterViewInit> {
+      if (tracingSpan) {
+        tracingSpan.finish();
+      }
+      if (originalAfterViewInit) {
+        // tslint:disable-next-line:no-unsafe-any
+        return originalAfterViewInit.apply(this, args);
+      }
+    };
+  };
+}
+
+/**
+ * Decorator function that can be used to capture a single lifecycle methods of the component.
+ */
+export function TraceMethodDecorator(): MethodDecorator {
+  return (target: Object, propertyKey: string | symbol, descriptor: PropertyDescriptor) => {
+    const originalMethod = descriptor.value;
+    descriptor.value = function(...args: any[]): ReturnType<typeof originalMethod> {
+      const now = timestampWithMs();
+      const activeTransaction = getActiveTransaction();
+      if (activeTransaction) {
+        activeTransaction.startChild({
+          description: `<${target.constructor.name}>`,
+          endTimestamp: now,
+          op: `angular.${String(propertyKey)}`,
+          startTimestamp: now,
```

Why is the start and end times the same here? Would it make sense to set the end time after the original method returns?

---

I did it like that because I didn't want to wrap method in a try/catch, as it'd swallow it from the original angular error flow. 99,9% method call itself will be almost instant anyway, so there's not enough to gain from measuring it, thus it's a point-in-time marker.